### ISSUE TITLE
ERROR:root:TraCI has disconnected with: 'NoneType' object has no attribute 'setOrder'

### ISSUE THREAD

**Issue by yuant95** [huawei-noah/SMARTS#1759]
**High Level Description**
After running the RL training for 19 hours. It started to get error `connect to TraCI server at localhost:54319 [Errno 111] Connection refused`. I'm not sure whether this is more related to SUMO? This happened while running multiple jobs on parallel using the same python environment. 

**Desired SMARTS version**
0.6.1

**Operating System**
ubuntu 20.04

**Problems**
Here is the traceback massege.

```
ERROR:root:Failed to initialize SUMO
                Your scenario might not be configured correctly or
                you were trying to initialize many SUMO instances at
                once and we were not able to assign unique port
                numbers to all SUMO processes.
                Check /ubc/cs/home/y/yuant426/.smarts/_sumo_run_logs/sumo-61a988b8 for hints
ERROR:root:TraCI has disconnected with: 'NoneType' object has no attribute 'setOrder'
WARNING:SumoTrafficSimulation:attempting to transfer SUMO vehicles to other providers...
ERROR:SMARTS:Failed to successfully reset after 2 times.
Traceback (most recent call last):
  File "competition/track1/train/run.py", line 281, in <module>
    main(args)
  File "competition/track1/train/run.py", line 95, in main
    run(envs_train=envs_train, envs_eval=envs_eval, config=config, wandb_run = wandb_run)
  File "competition/track1/train/run.py", line 142, in run
    model.learn(
  File "/ubc/cs/research/plai-scratch/smarts/miniconda3/envs/smartsEnv/lib/python3.8/site-packages/stable_baselines3/ppo/ppo.py", line 299, in learn
    return super(PPO, self).learn(
  File "/ubc/cs/research/plai-scratch/smarts/miniconda3/envs/smartsEnv/lib/python3.8/site-packages/stable_baselines3/common/on_policy_algorithm.py", line 250, in learn
    continue_training = self.collect_rollouts(self.env, callback, self.rollout_buffer, n_rollout_steps=self.n_steps)
  File "/ubc/cs/research/plai-scratch/smarts/miniconda3/envs/smartsEnv/lib/python3.8/site-packages/stable_baselines3/common/on_policy_algorithm.py", line 184, in collect_rollouts
    if callback.on_step() is False:
  File "/ubc/cs/research/plai-scratch/smarts/miniconda3/envs/smartsEnv/lib/python3.8/site-packages/stable_baselines3/common/callbacks.py", line 88, in on_step
    return self._on_step()
  File "/ubc/cs/research/plai-scratch/smarts/miniconda3/envs/smartsEnv/lib/python3.8/site-packages/stable_baselines3/common/callbacks.py", line 192, in _on_step
    continue_training = callback.on_step() and continue_training
  File "/ubc/cs/research/plai-scratch/smarts/miniconda3/envs/smartsEnv/lib/python3.8/site-packages/stable_baselines3/common/callbacks.py", line 88, in on_step
    return self._on_step()
  File "/ubc/cs/research/plai-scratch/smarts/SMARTS_VCR/competition/track1/train/wandbCallback.py", line 133, in _on_step
    episode_rewards, episode_lengths, episode_infos = evaluate_policy(
```

**Comment by Gamenot** [comment]
Sorry for the late reply on this, the error comes from the cause that SUMO can have a race condition for ports when acquiring a port if many SUMO instances are generated simultaneously. We generally saw this as solved with a particularly tiny possibility because it has to fail a port acquisition 15 times.

The failure cases are as follows:
- Connection to the SUMO process took longer than `5.05s` 
    - The SUMO process died
    - The SUMO process did not handshake in time (CPU lockup could do this)
- `traci` attempted to connect to a port that is in use
   - `traci` tried to connect to another SUMO instance
   - `traci` tried to use a port that some other program is using

We struggle with this a bit since it is hard to guarantee SUMO since it can crash for legitimate reasons such as a malformed resource file.

**Comment by Gamenot** [comment]
I think there is still a bit more room to improve in handling these cases since we can perhaps check the exit-code of the SUMO process to see if it is still viable aside from the viability of the instance from port connection response.

@saulfield

**Comment by yuant95** [comment]
I am actaully running 3 RL agent training on parallel, and this error kills all of them at the same time after 19 hours. I'm guessing this is indicating that running SUMO server on parallel may be a issue?

**Comment by Gamenot** [comment]
@yuant95 Would there happen to be stranded processes left over? It is also possible that SMARTS or the framework is not cleaning up the processes at the end.

If your model is very intensive it is just possible that SMARTS cannot afford to keep a timeout. I will look into adding in configuration to attempt retries until it succeeds. 

I am also going to try using a stress test to see what might be up with it.

**Comment by Gamenot** [comment]
The newly merged #1776 should prevent this issue from occurring.
